syntax = "proto2";

package core;

message Message {

    enum Type {
        WORKFLOWPACKET = 1;
        WORKFLOWPACKETDATA = 2;
        ACTIONPACKET = 3;
        ACTIONPACKETDATA = 4;
        GENERALPACKET = 5;
    }

    optional Type type = 1;
    optional WorkflowPacket workflow_packet = 2;
    optional WorkflowPacketData workflow_packet_data = 3;
    optional ActionPacket action_packet = 4;
    optional ActionPacketData action_packet_data = 5;
    optional GeneralPacket general_packet = 6;

}

message WorkflowPacket {

    message WorkflowSender {
        optional string name = 1;
        optional string uid = 2;
        optional string workflow_execution_uid = 3;
    }

    optional WorkflowSender sender = 1;
    optional string callback_name = 2;
}

message WorkflowPacketData {
    optional WorkflowPacket.WorkflowSender sender = 1;
    optional string callback_name = 2;
    optional string additional_data = 3;
}

message ActionPacket {

    message ActionArgument {
        optional string name = 1;
        optional string value = 2;
        optional string reference = 3;
        optional string selection = 4;
    }

    message ActionSender {
        optional string name = 1;
        optional string uid = 2;
        optional string execution_uid = 3;
        optional string app_name = 4;
        optional string action_name = 5;
        repeated ActionArgument arguments = 6;
        optional string workflow_execution_uid = 7;
    }

    optional ActionSender sender = 1;
    optional string callback_name = 2;
}

message ActionPacketData {

    optional ActionPacket.ActionSender sender = 1;
    optional string callback_name = 2;
    optional string additional_data = 3;
}

message GeneralPacket {

    message GeneralSender {
        optional string uid = 1;
        optional string app_name = 2;
        optional string workflow_execution_uid = 3;
    }

    optional GeneralSender sender = 1;
    optional string callback_name = 2;
}
